### ISSUE TITLE
Frame.insert_before

### ISSUE THREAD

**Issue by ForeverWintr** [InvestmentSystems/static-frame#213]
Although inserting a column in the middle or at the start of a Frame is undesirable for performance reasons, it is sometimes unavoidable. I find myself in this situation semi often when needing to add informational columns to the front of a frame, for example. Usually I accomplish this by

1. Converting the `Frame` to a `FrameGO`
2. Inserting the new column[s] at the end
3. reorganizing the columns using something like the following:
```python
desired_order = ['new_column'] + [c for c in frame.columns if c != 'new_column']
frame = frame.reindex(columns=desired_order)
```

Exposing a convenience method on `Frame` would make this easier, and possibly more performant. I suggest `insert_before`, which takes the name of an existing column in the frame, and a series (or frame) to insert. For example, given Frame `f`:

```python
f
<FrameGO>
<IndexGO> a       c       <<U1>
<Index>
0         1       4
1         2       5
2         3       6
<int64>   <int64> <int64>
```

A new column could be inserted like so:

```python
f.insert_before('c', sf.Series([1, 2], name='b'))
f
<FrameGO>
<IndexGO> a       b         c       <<U1>
<Index>
0         1       1.0       4
1         2       2.0       5
2         3       nan       6
<int64>   <int64> <float64> <int64>
```

## Comparison to pandas

By contrast, [pandas insert](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.insert.html) takes 4 arguments:

```
loc: int
    Insertion index. Must verify 0 <= loc <= len(columns).

column: str, number, or hashable object
    Label of the inserted column.

value: int, Series, or array-like

allow_duplicates: bool, optional
```

I see some drawbacks to this:

1. It's not clear what happens to the existing column at `loc`. Is it replaced by the inserted column? Moved left/right? (the answer seems to be 'moved left', based on my testing).
    - by contrast, `insert_before` makes this clear.
2. `insert` operates in place.
3. `insert` uses positional index (which may be necessary when columns are not guaranteed unique). This means if you want to insert by label, you must first find its positional index (an `O(n)` operation).
    - By contrast, finding the label at a given positional index is constant time, and those wishing to use `insert_before` with a positional index can just do `f.insert_before(f.columns[2], sf.Series([1, 2]))`.

One other difference is that `insert` requires you to specify a name, and ignores the name on named series you pass. Although this is consistent with the behavior of `__setitem__`, I think it's simpler to simply use the name on the provided series.

**Comment by flexatone** [comment]
Many thanks for this suggestion.

Overall, I like the interface as suggested. However, I think the name of the method might simply be `insert`, as we have precedent in the Python list, always have unique labels, and already follow list-like operations for `append` and `extend` on the columns.

**Comment by ForeverWintr** [comment]
I think `insert` makes sense for position based insertion as in pandas (and `list.insert`), but not for key based insertion, which is what I'm proposing here. 

When I do `mylist.insert(3, element)` I am expecting to insert `element` at position 3. I can then retrieve my element by indexing: `mylist[3] == element`. In concept, I am changing the item available at position 3.

However, unlike lists, `Frame`s are generally indexed by column name, not positional index. If I say `f.insert('c', series)`, I'm *not* inserting `series` at position `'c'`. `f['c']` still gives me the original `'c'` column (`f['c'] is not series`). I am not changing the item available at key `'c'`.

I think this difference is enough to justify a different interface, and shoehorning this behavior into an `insert` interface will likely be confusing. If we do go with `insert`, I think it's better to operate positionally as pandas does. I'm slightly opposed to that because in practice we usually operate with column names and the translation from column name to position is somewhat cumbersome.

**Comment by flexatone** [comment]
Thanks for your further thoughts.

I see your point: in the case of a `list`, `insert(x, y)` puts `y` at position `x`; whereas on a `Frame` using a `loc` label, `insert(x, y)`  puts `y` in the `iloc` position formerly held by the label and values of `x`; `x` will still be associated with the same values.

`Frame`s are indexed by `loc` labels and `iloc` labels simultaneously and in a 1-to-1 manner. It is common that operations can change one of those relationships and leave the other unchanged. For example, reindexing by the reverse of an index changes the `iloc` associations but retains the `loc` associations.

In the case of this proposed `insert`, we are inserting at a stable `iloc` position but changing the `loc` relationship. I understand that it is different than how a `list` works, but I think it is in-line with the spirit of many operations where we have both `iloc` and `loc` representations at the same time. 

That said, I am open to a different name. If we have `insert_before` it begs for an `insert_after`; would it be useful to have both? If so, I am more inclined toward that naming approach.

**Comment by ForeverWintr** [comment]
Yes, the `iloc` position we are inserting at remains stable, but unless we're using the iloc interface I think it's confusing to use metaphors that apply to it. 

If I see `frame.insert('c', sf.Series([1, 2]))` I would expect the series to be inserted at 'c'. Wouldn't you?

I don't see a need for `insert_after`, unless we want to expose it as an alternative for appending columns without going to a FrameGO, and I'm not aware of a use case that calls for that. By providing `insert_before` but not `insert_after`, we're explicitly saying that we allow inserting before the target column only, and I think that's clear. 

We could consider alternate names too, like `insert_left` or `linsert` (like `str.lstrip` or `str.ljust`).

**Comment by nekobon** [comment]
I think if we have `insert_left`, we would also want `insert_right`. It is true that for appending at the end, it doesn't make sense to use `insert_right`. But when we use it to add a new column in the middle, it seems strange we can use the right column as an anchor but not the one before.

If the goal is to allow inserting to the front of the Frame, perhaps it is better to have an explicit `insert_front` or `append_front` that only adds a column to the left-end. If we want to support inserting to anywhere, I would expect there is `insert_after` if there is `insert_before`.

**Comment by flexatone** [comment]
Thanks again, @ForeverWintr , for this suggestion: this method, as well as an `insert_after`, is now implemented on `Frame` (and `Series`). In addition to your original suggestion, we have added a `fill_value` argument, as well as support for inserting an entire `Frame`.

This feature will be available in the 0.6.18 release.


```python
>>> f = sf.Frame(np.arange(6).reshape(3, 2), columns=('a', 'c'))
>>> f
<Frame>
<Index> a       c       <<U1>
<Index>
0       0       1
1       2       3
2       4       5
<int64> <int64> <int64>

>>> f.insert_before('c', sf.Series([1, 2], name='b'))
<Frame>
<Index> a       b         c       <<U1>
<Index>
0       0       1.0       1
1       2       2.0       3
2       4       nan       5
<int64> <int64> <float64> <int64>

>>> f.insert_before('c', sf.Series([1, 2], name='b'), fill_value=0)
<Frame>
<Index> a       b       c       <<U1>
<Index>
0       0       1       1
1       2       2       3
2       4       0       5
<int64> <int64> <int64> <int64>

>>> f.insert_before('c', f.relabel(columns=lambda x: x.upper()))
<Frame>
<Index> a       A       C       c       <<U1>
<Index>
0       0       0       1       1
1       2       2       3       3
2       4       4       5       5
<int64> <int64> <int64> <int64> <int64>
```
